doc/dev/crimson/pipelines.rst

```diff
@@ -0,0 +1,258 @@
+=========
+Pipelines
+=========
+
+RADOS Protocol Requirements
+===========================
+
+Requests
+--------
+
+OSD serves different kinds of requests. Some of them are targeting the local state of the connected
+OSD instance. For instance, the ``ceph config get`` command is directed straight to the local config
+subsystem, and is fulfilled right away. But others inquiry or mutate the shared state of the cluster.
+Following are two categories of these requests:
+
+- I/O requests from clients.
+- PG peering requests from peer OSDs.
+
+Because inherently a distributed storage system is distributed by its own nature, we can not assume
+that an OSD or the PGs hosted by it are always ready to serve a request targeting a shared state, not
+to mention the unpredictable hardware failures. But fortunately, even if an OSD is not able to serve
+a request right away, under most circumstances, it can postpone the request, prepare itself so that it
+can answer this request. It's quite common in daily life:
+
+  Great question, let me look into it and I will get back to you!
+
+Some questions involves multiple steps. To serve an I/O request from client, at
+least 7 stages are involved:
+
+.. ditaa::
+
+   +----+---------------------+
+   |cGRE| OSD wait for osdmap |
+   +----+----+----------------+
+             |
+             V
+   +----+-----------------+
+   |cGRE| OSD wait for PG |
+   +----+-----------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for osdmap |
+   +----+--------------------+
+             |
+             V
+   +----+--------------------+
+   |cBLU| PG wait for active |
+   +----+--------------------+
+             |
+             V
+   +----+----------------------+
+   |cBLU| PG wait for recovery |
+   +----+----------------------+
+             |
+             V
+   +----+---------------------+
+   |cBLU| wait for object ctx |
+   +----+---------------------+
+             |
+             V
+   +----+------------------+
+   |cBLU| wait for process |
+   +----+------------------+
+
+             legend
+             +----------------+
+             |cGRE OSD stages |
+             +----------------+
+             |cBLU PG stage   |
+             +----------------+
+
+Two Requirements
+----------------
+
+An OSD client talks with each connected OSD over stateful connection. The client is allowed to
```

Comment:
It's worth a sentence explaining that stateful here means that all messages will be delivered in order without missing messages.  An OSD's messenger will continue trying to reconnect to a peer osd to which it's connection has failed (holding any unconfirmed messages in memory) until either it reconnects or it receives an OSDMap from someone else (or a monitor) marking the peer OSD down.  You could skip this section on mechanism by replacing this sentence with "An OSD talks with each peer OSD over a connection which guarantees in order message delivery unless marked down in the OSDMap".  Also, s/OSD client/OSD since clients use lossy connections which can drop messages -- the Objecter will resend ops in that case after reconnect independently of the messenger.

Reply:
thanks. incorporated.